controllers/utils/multicloud_util.go

```diff
@@ -0,0 +1,341 @@
+		if labels == nil {
+			labels = make(map[string]string)
+		}
+		labels["openshift-cluster-group-upgrades/clusterGroupUpgrade"] = cguLabel
+		mcv.SetLabels(labels)
+
+		viewSpec := viewv1beta1.ViewSpec{
+			Scope: viewv1beta1.ViewScope{
+				Resource:  resourceType,
+				Name:      resourceName,
+				Namespace: resourceNamespace,
+			},
+		}
+
+		mcv.Spec = viewSpec
+		if err := c.Update(ctx, mcv); err != nil {
+			return nil, err
+		}
+	}
+	// Give some time for the ManagedClusterView to retrieve the info.
+	time.Sleep(ManagedClusterActionWaitTimeSec * time.Second)
+	return mcv, nil
+}
+
+// EnsureManagedClusterActionForInstallPlan creates or updates an action for an InstallPlan.
+func EnsureManagedClusterActionForInstallPlan(
+	ctx context.Context, c client.Client, name string, namespace string,
+	installPlan operatorsv1alpha1.InstallPlan, Log logr.Logger) (*actionv1beta1.ManagedClusterAction, error) {
+
+	mcaForInstallPlan := &actionv1beta1.ManagedClusterAction{}
+	if err := c.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, mcaForInstallPlan); err != nil {
+		// If the specific managedClusterAction was not found, create it.
+		if errors.IsNotFound(err) {
+			Log.Info("[EnsureManagedClusterActionForInstallPlan] MCA doesn't exist, create it",
+				"name", name, "namespace", namespace)
+			actionMeta := metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			}
+			actionSpec, err := NewManagedClusterActionForInstallPlanSpec(installPlan)
+			if err != nil {
+				return nil, err
+			}
+			mcaForInstallPlan = &actionv1beta1.ManagedClusterAction{
+				ObjectMeta: actionMeta,
+				Spec:       actionSpec,
+			}
+
+			if err := c.Create(ctx, mcaForInstallPlan); err != nil {
+				return mcaForInstallPlan, nil
+			}
+		}
+	} else {
+		// If the specific managedClusterView was found, update it.
+		Log.Info("[EnsureManagedClusterActionForInstallPlan] MCA already exists, update it",
+			"name", name, "namespace", namespace)
+		actionSpec, err := NewManagedClusterActionForInstallPlanSpec(installPlan)
+		if err != nil {
+			return nil, err
+		}
+		mcaForInstallPlan.Spec = actionSpec
+
+		if err := c.Update(ctx, mcaForInstallPlan); err != nil {
+			return nil, err
+		}
+	}
+
+	return mcaForInstallPlan, nil
+}
+
+// NewManagedClusterActionForInstallPlanSpec returns the action spec for approving an InstallPlan.
+func NewManagedClusterActionForInstallPlanSpec(installPlan operatorsv1alpha1.InstallPlan) (actionv1beta1.ActionSpec, error) {
+	installPlanClusterServiceVersionNames, err := json.Marshal(installPlan.Spec.ClusterServiceVersionNames)
+	if err != nil {
+		fmt.Printf("Error trying to Marshal InstallPlan ClusterServiceVersionNames: %s", err.Error())
+	}
+
+	templateContent := fmt.Sprintf(
+		`{"apiVersion": "operators.coreos.com/v1alpha1","kind": "InstallPlan",
+		  "metadata": {"name": "%s","resourceVersion": "%s"},
+		  "spec": {"approval": "Manual","approved": true, "clusterServiceVersionNames": %s}}`,
```

Comment:
This should get applied as a patch in which case I don't think we want to include the "approval: Manual" field (unless we have to). Including it means that any other logic error in the code could lead to us changing the approval type of an installplan (obviously without meaning to). Same thing for the CSV version, we would not want to overwrite/update that, it should already be set to the right value.

Reply:
This is not how actions work, unfortunately. Even though I set the action type to update: "ActionType: actionv1beta1.UpdateActionType" (in NewManagedClusterActionForInstallPlanSpec), these fields are still required, otherwise you can't apply them, including the CSV name - this is the reason I had to create the managedClusterView for the installPlan, to get access to the CSV and resource version. Also, that is why I made the checks above to make sure the installPlan has the approval set to Manual.

@serngawy had the same surprise when he discovered we have to include the resource version. He can comment/conform/share more options and I will triple check once I can recover my setup.

Reply:
The resourceVersion I kind of understand (helps to uniquely identify the target of the action perhaps?), but the requirement to include fields that we don't intend (or want) to modify is strange. We should raise this with the ACM team to understand why we can't patch just one field without specifying the others.

Reply:
I did followup on a discussion Mohamed had with ACM on this same topic. Their answer to Mohamed's question "why do we also need resourceVersion" was "Currently, I think it’s necessary for some resources. But I think we could investigate if it’s possible to  enhance it in our code", so it doesn't seem the same for all the resources. We'll see what they say, but for now we need all the fields in order for this to work.